base/src/commonMain/kotlin/utils/timer/MonotonicTimer.kt

```diff
@@ -0,0 +1,145 @@
+            }
+        }
+    }
+
+    override suspend fun pause() {
+        takeAndChangeState { state ->
+            when (state) {
+                is State.Running -> state::pause
+                is State.NotRunning -> state.remain()
+            }
+        }
+    }
+
+    private suspend fun finish() {
+        takeAndChangeState { state ->
+            when (state) {
+                is State.NotRunning.Finished -> state.remain()
+                is State.Running -> state::finish
+                is State.NotRunning.Paused -> state::finish
+            }
+        }
+    }
+
+    /** Timer state. */
+    sealed class State : KalugaState(), Timer.State {
+        abstract val totalDuration: Duration
+        /** Timer is not running. */
+        sealed class NotRunning(override val elapsed: Duration) : State(), Timer.State.NotRunning {
+            /** Timer is paused. */
+            class Paused(
+                elapsedSoFar: Duration,
+                override val totalDuration: Duration
+            ) : NotRunning(elapsedSoFar), Timer.State.NotRunning.Paused {
+                fun start(
+                    coroutineScope: CoroutineScope,
+                    finishCallback: suspend () -> Unit
+                ): Running = Running(elapsed, totalDuration, coroutineScope, finishCallback)
+                internal fun finish(): Finished = Finished(totalDuration)
+            }
+
+            /** Timer is finished. */
+            class Finished(
+                override val totalDuration: Duration
+            ) : NotRunning(totalDuration), Timer.State.NotRunning.Finished
+        }
+
+        /** Timer is running. */
+        class Running(
+            elapsedSoFar: Duration,
+            override val totalDuration: Duration,
+            private val coroutineScope: CoroutineScope,
+            private val finishCallback: suspend () -> Unit
+        ) : State(), Timer.State.Running, HandleAfterNewStateIsSet<State>, HandleBeforeOldStateIsRemoved<State> {
+            override val elapsed: Flow<Duration> = tickProvider(elapsedSoFar, totalDuration)
+            private lateinit var finishJob: Job
+
+            internal suspend fun pause(): NotRunning.Paused {
+                return NotRunning.Paused(elapsed.first(), totalDuration)
+            }
+            internal fun finish(): NotRunning.Finished = NotRunning.Finished(totalDuration)
+
+            override suspend fun beforeOldStateIsRemoved(oldState: State) {
+                finishJob = coroutineScope.launch {
+                    delay(totalDuration - elapsed.first())
+                    finishCallback()
+                }
+            }
+            override suspend fun afterNewStateIsSet(newState: State) {
+                finishJob.cancel()
+            }
+        }
+    }
+}
+
+private fun tickProvider(offset: Duration, max: Duration): Flow<Duration> {
+    val mark = TimeSource.Monotonic.markNow()
+
+    return flow {
```

Comment:
Does this just loop without any delay?

Reply:
yes, the idea was that the consumer controls the ticks. I could create an extension function something like
```
fun Timer.elapsedWithInterval(interval: Duration) = elapsed.map { it.also { delay (interval) }}
```

Reply:
I think the regular user for this will, for example in their viewmodel, just observe the elapsed on running, leading to a _lot_ of updates.

So I think a method to get a flow with interval should be exposed there, and if a non interval flow is offered then it should be explicit that this is the behaviour in the method/property.

To further help the user the interval should express it is a running target (your proposed method will not produce regular ticks) and undershoot by default (controlled by a flag perhaps) as this is it a common mistake to rely on this as accurate.

Either that or make something that will produce the exact intervals as values as close as possible to the correct time (which would be quite nice)

Reply:
implemented 3 options: see `TickCorrection`

Reply:
To see how they work either text or documentation in README is needed to evaluate them.

Reply:
when playing around with timers, found that one approach is better than the other. cleaning up the code and updating the readme

Reply:
Good idea to keep the best ones only.